src/pyhf/contrib/viz/brazil.py

```diff
@@ -24,27 +69,212 @@ def plot_results(ax, mutests, tests, test_size=0.05):
+            mutests,
+            cls_exp[idx],
+            color=color,
+            linestyle="dotted" if idx != 2 else "dashed",
+            label=None if idx != 2 else r"$\mathrm{CL}_{s,\mathrm{exp}}$",
         )
-    ax.fill_between(mutests, cls_exp[0], cls_exp[-1], facecolor='yellow')
-    ax.fill_between(mutests, cls_exp[1], cls_exp[-2], facecolor='green')
-    ax.plot(mutests, [test_size] * len(mutests), c='red')
+        cls_exp_line_artists.append(_cls_exp_line)
+    one_sigma_band = ax.fill_between(
+        mutests,
+        cls_exp[0],
+        cls_exp[-1],
+        facecolor="yellow",
+        label=r"$\pm2\sigma$ $\mathrm{CL}_{s,\mathrm{exp}}$",
+    )
+    two_sigma_band = ax.fill_between(
+        mutests,
+        cls_exp[1],
+        cls_exp[-2],
+        facecolor="green",
+        label=r"$\pm1\sigma$ $\mathrm{CL}_{s,\mathrm{exp}}$",
+    )
+
+    test_size_color = kwargs.pop("test_size_color", "red")
+    test_size_linestyle = kwargs.pop("test_size_linestyle", "solid")
+    (test_size_line,) = ax.plot(
+        mutests,
+        [test_size] * len(mutests),
+        color=test_size_color,
+        linestyle=test_size_linestyle,
+        label=rf"$\alpha={test_size}$",
+    )
     ax.set_ylim(0, 1)
 
-    ax.set_xlabel(r"$\mu$ (POI)")
-    ax.set_ylabel(r"$\mathrm{CL}_{s}$")
+    x_label = kwargs.pop("xlabel", r"$\mu$ (POI)")
+    y_label = kwargs.pop("ylabel", r"$\mathrm{CL}_{s}$")
+    ax.set_xlabel(x_label)
+    ax.set_ylabel(y_label)
+
+    return BrazilBandContainer(
+        (
+            cls_obs_line_artist,
+            cls_exp_line_artists,
+            [one_sigma_band, two_sigma_band],
+            test_size_line,
+        ),
+    )
+
+
+def plot_cls_components(
+    ax,
+    mutests,
+    tests,
+    test_size=0.05,
+    no_clb=False,
+    no_clsb=False,
+    no_cls=False,
+    **kwargs,
+):
+    r"""
+    Plot the values of :math:`\mathrm{CL}_{s+b}` and :math:`\mathrm{CL}_{b}`
+    --- the components of the :math:`\mathrm{CL}_{s}` ratio --- on top of the
+    :math:`\mathrm{CL}_{s}` values for a series of hypothesis tests for various
+    POI values.
+
+    Example:
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> import pyhf
+        >>> import pyhf.contrib.viz.brazil
+        >>> pyhf.set_backend("numpy")
+        >>> model = pyhf.simplemodels.hepdata_like(
+        ...     signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
+        ... )
+        >>> observations = [51, 48]
+        >>> data = observations + model.config.auxdata
+        >>> poi_vals = np.linspace(0, 5, 41)
+        >>> results = [
+        ...     pyhf.infer.hypotest(
+        ...         test_poi, data, model, return_expected_set=True, return_tail_probs=True
+        ...     )
+        ...     for test_poi in poi_vals
+        ... ]
+        >>> fig, ax = plt.subplots()
+        >>> artists = pyhf.contrib.viz.brazil.plot_cls_components(ax, poi_vals, results)
```

Comment:
why does this return multiple artists, oculdn't we add the components to the `BrazilBandContainer?`

Reply:
As `BrazilBandContainer` acts as a `namedtuple` then would the idea be to have `plot_results` have something like

```python
    return BrazilBandContainer(
        (
            cls_obs_line,
            cls_exp_lines,
            one_sigma_band,
            two_sigma_band,
            test_size_line,
            None,
            None
        ),
    )
```

then? That seems a little strange as now `plot_results` returns an object that has public attributes that will never get used by it. Not sure what's the best approach here.

Reply:
I think of it like default ctor args that are also often `None`

Reply:
this does not work?

```
    return BrazilBandContainer(
        (
            cls_obs_line,
            cls_exp_lines,
            one_sigma_band,
            two_sigma_band,
            test_size_line,
        ),
    )
```

and when we have it just pass a larger tuple?

```
    return BrazilBandContainer(
        (
            cls_obs_line,
            cls_exp_lines,
            one_sigma_band,
            two_sigma_band,
            test_size_line,
            pb_line,
            psb_line,
        ),
    )
```

Reply:
> this does not work?
> 
> ```
>     return BrazilBandContainer(
>         (
>             cls_obs_line,
>             cls_exp_lines,
>             one_sigma_band,
>             two_sigma_band,
>             test_size_line,
>         ),
>     )
> ```
> 
> and when we have it just pass a larger tuple?
> 
> ```
>     return BrazilBandContainer(
>         (
>             cls_obs_line,
>             cls_exp_lines,
>             one_sigma_band,
>             two_sigma_band,
>             test_size_line,
>             pb_line,
>             psb_line,
>         ),
>     )
> ```

I'll have to check later, but I'm pretty sure this doesn't work as `matplotlib.container.Container` should behave like `collections.namedtuple` (c.f. [this comment in `matplotlib.container.StemContainer`](https://matplotlib.org/stable/_modules/matplotlib/container.html#StemContainer)) and 

```python
>>> from collections import namedtuple
>>> Point = namedtuple('Point', ['x', 'y'])
>>> Point(11, 12, z=5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __new__() got an unexpected keyword argument 'z'
>>> Point(11, 12, 5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __new__() takes 3 positional arguments but 4 were given
```

but I might be wrong.

Reply:
@henryiii what are your views on having a `matplotlib.container.Container` that has `None` for some of the artists attributes, like described in the above?

Reply:
My current plan is to just start factorizing the plotting into smaller functions and then make more container classes: `BrazilBandContainer`, `ClsComponentsContainer`, `ResultsPlotContainer`

Reply:
To summarize some problems we've had with trying to reduce things down to fewer containers as Lukas was hoping for:

I don't think we want to do something like

```python
class BrazilBandContainer(Container):
    def __init__(self, brazil_band_artists, clsb=None, clb=None, **kwargs):
        (
            cls_obs,
            cls_exp,
            one_sigma_band,
            two_sigma_band,
            test_size,
        ) = brazil_band_artists
        self.cls_obs = cls_obs
        self.cls_exp = cls_exp
        self.one_sigma_band = one_sigma_band
        self.two_sigma_band = two_sigma_band
        self.test_size = test_size
        self.clsb = clsb
        self.clb = clb
        super().__init__(brazil_band_artists, **kwargs)
```

as since `clsb` and `clb` do not end up in the `namedtuple`-like container (but just as attributes) and so you have them as strangely tacked on artists that are kept separate from the rest of the artists. I also don't think that you want to go with

```python
class BrazilBandContainer(Container):
    def __init__(self, brazil_band_artists **kwargs):
        (
            cls_obs,
            cls_exp,
            one_sigma_band,
            two_sigma_band,
            test_size,
            clsb,
            clb
        ) = brazil_band_artists
        self.cls_obs = cls_obs
        self.cls_exp = cls_exp
        self.one_sigma_band = one_sigma_band
        self.two_sigma_band = two_sigma_band
        self.test_size = test_size
        self.clsb = clsb
        self.clb = clb
        super().__init__(brazil_band_artists, **kwargs)
```

as then in `plot_brazil_band()` you're returning

```python
    return BrazilBandContainer(
        (
            cls_obs_line,
            cls_exp_lines,
            one_sigma_band,
            two_sigma_band,
            test_size_line,
            None,
            None,
        )
    )
```

which is a bit awkward to be forced to tack on the `None`s, and then you need to unpack and repack in `plot_results()` to combine this with a `ClsComponentsContainer` if you want to return a `BrazilBandContainer`, which feels like forcing things into a container that don't need to be.

I'm not sure what's best to do, but It feels like this is trying to combine too many things into a single container, as the [`container` classes in `matplotlib`](https://matplotlib.org/stable/api/container_api.html) all only are used for one group of artists plotted on a subplot.